classic/ethhash.go

```diff
@@ -0,0 +1,618 @@
+		log.Info("Disk storage enabled for ethash DAGs", "dir", config.DatasetDir, "count", config.DatasetsOnDisk)
+	}
+	ethash := &Ethash{
+		config:   config,
+		caches:   newlru(config.CachesInMem, newCache),
+		datasets: newlru(config.DatasetsInMem, newDataset),
+		update:   make(chan struct{}),
+		// hashrate: metrics.NewMeterForced(),
+	}
+	if config.PowMode == ModeShared {
+		ethash.shared = sharedEthash
+	}
+	ethash.remote = startRemoteSealer(ethash, notify, noverify)
+	return ethash
+}
+
+func (ethash *Ethash) Threads() int {
+	ethash.lock.Lock()
+	defer ethash.lock.Unlock()
+
+	return ethash.threads
+}
+
+// verifyHeader checks whether a header conforms to the consensus rules of the
+// stock Ethereum ethash engine.
+// See YP section 4.3.4. "Block Header Validity"
+func (ethash *Ethash) verifyHeader(chain ChainHeaderReader, header, parent *types.Header, uncle bool, seal bool, unixNow int64) error {
+	// Ensure that the header's extra-data section is of a reasonable size
+	if uint64(len(header.Extra)) > MaximumExtraDataSize {
+		return fmt.Errorf("extra-data too long: %d > %d", len(header.Extra), MaximumExtraDataSize)
+	}
+	// Verify the header's timestamp
+	if !uncle {
+		if header.Time > uint64(unixNow+int64(allowedFutureBlockTime.Seconds())) {
+			return ErrFutureBlock
+		}
+	}
+	if header.Time <= parent.Time {
+		return errOlderBlockTime
+	}
+	// Verify the block's difficulty based on its timestamp and parent's difficulty
+	expected := ethash.CalcDifficulty(chain, header.Time, parent)
+
+	if expected.Cmp(header.Difficulty) != 0 {
+		return fmt.Errorf("invalid difficulty: have %v, want %v", header.Difficulty, expected)
+	}
+	// Verify that the gas limit is <= 2^63-1
+	if header.GasLimit > MaxGasLimit {
+		return fmt.Errorf("invalid gasLimit: have %v, max %v", header.GasLimit, MaxGasLimit)
+	}
+	// Verify that the gasUsed is <= gasLimit
+	if header.GasUsed > header.GasLimit {
+		return fmt.Errorf("invalid gasUsed: have %d, gasLimit %d", header.GasUsed, header.GasLimit)
+	}
+	// Verify the block's gas usage and (if applicable) verify the base fee.
+	if !ethash.pluginConfig.IsEnabled(ethash.pluginConfig.GetEIP1559Transition, header.Number) {
+		// Verify BaseFee not present before EIP-1559 fork.
+		if header.BaseFee != nil {
+			return fmt.Errorf("invalid baseFee before fork: have %d, expected 'nil'", header.BaseFee)
+		}
+		if err := VerifyGaslimit(parent.GasLimit, header.GasLimit); err != nil {
+			return err
+		}
+	} else if err := VerifyEIP1559Header(*ethash.pluginConfig, parent, header); err != nil {
+		// Verify the header's EIP-1559 attributes.
+		return err
+	}
+	// Verify that the block number is parent's +1
+	if diff := new(big.Int).Sub(header.Number, parent.Number); diff.Cmp(big.NewInt(1)) != 0 {
+		return ErrInvalidNumber
+	}
+	if ethash.pluginConfig.IsEnabledByTime(ethash.pluginConfig.GetEIP3860TransitionTime, &header.Time) || ethash.pluginConfig.IsEnabled(ethash.pluginConfig.GetEIP3860Transition, header.Number) {
```

Comment:
This whole if statement (and the one that follows) seem totally unnecessary, given that we're checking the plugin's config provided by the plugin against a chain config provided by the plugin to error out rather than support a feature set ETC may one day actually support.